[weiliann] iP

weiliann opened the pull request:
## JoeBot

> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes))

JoeBot frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~Fast~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😄 

And it is **FREE!**

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

weiliann committed:
Rename ChatBot to Joe

weiliann committed:
Add echo function to ChatBot

weiliann committed:
Allow ChatBot to save tasks in memory

weiliann committed:
Add Task class

weiliann committed:
Add ability to toggle task status

weiliann committed:
Add support for tracking three types of tasks

weiliann committed:
Add semi-automated Text UI testing

weiliann committed:
Allow ChatBot to deal with incorrect user inputs

weiliann committed:
Add support for deleting tasks from the list

weiliann committed:
Add support for saving tasks into local file

weiliann committed:
Merge branch 'branch-Level-7'

weiliann committed:
Require a valid date time to be input and use more oop

weiliann committed:
Merge branch 'branch-Level-8'

weiliann committed:
Organise classes into packages

weiliann committed:
Merge branch 'add-gradle-support'

weiliann committed:
Add test for parser class

weiliann committed:
Merge branch 'add-junit-test'

weiliann committed:
Add some java documentation for the codebase

weiliann committed:
Tweak code to comply with coding standard

weiliann committed:
Add support to find a task with a keyword in the description

weiliann committed:
Merge branch 'A-JavaDoc'

weiliann committed:
Merge branch 'A-CodingStandard'

weiliann committed:
Merge branch 'Level-9'

weiliann committed:
Refactor the code to use more OOP principles

weiliann committed:
Add tests for Parser

weiliann committed:
Refactor code style

weiliann committed:
Refactor javadoc for codebase

weiliann committed:
Refactor codebase to adhere to checkstyle

weiliann committed:
Add GUI for joe

weiliann committed:
Merge branch 'branch-Level-10'

stephaneleong commented on src/main/java/joe/Joe.java:
```diff
@@ -0,0 +1,88 @@
+            try {
+                String fullCommand = ui.readCommand();
+                ui.printDivider();
+                Command c = Parser.parse(fullCommand);
```
Perhaps a more intuitive variable name here? may seem like a scratch variable

stephaneleong commented on src/main/java/joe/Storage.java:
```diff
@@ -0,0 +1,93 @@
+import joe.task.Event;
+import joe.task.Task;
+import joe.task.TaskList;
+import joe.task.Todo;
```
Good Job on organising your import statements 👍🏽

stephaneleong commented on src/main/java/joe/command/ExitCommand.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    @Override
+    public boolean isExit() {
```
Well Done on naming your boolean method appropriately 😄

stephaneleong commented on src/main/java/joe/task/Task.java:
```diff
@@ -0,0 +1,101 @@
+     * @param isDone The new completion status of the task.
+     * @return The updated {@code Task} object.
+     */
+    public Task setIsDone(boolean isDone) {
```
Perhaps you can rename this setter method to 'setDone'

stephaneleong commented on src/main/java/joe/task/Todo.java:
```diff
@@ -0,0 +1,50 @@
+
+    /**
+     * Serializes the {@code Todo} task into a formatted string to be stored in a file.
+     * The format is "T|isDone|description".
```
Awesome idea to put the format in the javadocs! 💯

stephaneleong commented on src/test/java/joe/ParserTest.java:
Good Job on doing thorough unit testing 🥇 ! I would suggest using underscores in your naming of the test methods (e.g. featureUnderTest_testScenario_expectedBehavior())

stephaneleong reviewed (commented):
LGTM! Did an overal great job in keeping to the Java Coding Standard. Just some suggestions to look through

ahmadsyuaib commented on src/main/java/joe/Parser.java:
```diff
@@ -0,0 +1,118 @@
+                throw new JoeException("Oops! Try: deadline {desc} /by {duedate}");
+            }
+            String taskDesc = String.join(" ", Arrays.copyOfRange(inputArr, 1, byIdx));
+            String taskBy = String.join("", Arrays.copyOfRange(inputArr, byIdx + 1, inputArr.length));
```
Is taskBy a good name? Perhaps a more descriptive name that includes a noun such as taskDueDate.

ahmadsyuaib commented on src/main/java/joe/Storage.java:
```diff
@@ -0,0 +1,93 @@
+            String line = null;
+            while ((line = reader.readLine()) != null) {
+                String[] lineArr = line.split("\\|");
+                String taskType = lineArr[0];
+                boolean isDone = Boolean.parseBoolean(lineArr[1]);
+                String taskDesc = lineArr[2];
```
Proper naming has been done here! Good job!

ahmadsyuaib commented on src/main/java/joe/task/TaskList.java:
```diff
@@ -0,0 +1,114 @@
+     * @param query The query string to search for within the task descriptions.
+     * @return An {@code ArrayList} of tasks that match the query.
+     */
+    public ArrayList<Task> find(String query) {
```
Is find descriptive enough?

ahmadsyuaib commented on src/main/java/joe/task/TaskList.java:
```diff
@@ -0,0 +1,114 @@
+     *
+     * @return The {@code ArrayList} containing all the tasks.
+     */
+    public ArrayList<Task> toArrayList() {
+        return this.taskList;
+    }
```
Would a getter method fit better here since your are returning this.taskList directly? Perhaps getArrayList()?

ahmadsyuaib commented on src/main/java/joe/task/TaskList.java:
```diff
@@ -0,0 +1,114 @@
+     *
+     * @return {@code true} if the task list is empty, {@code false} otherwise.
+     */
+    public boolean isEmpty() {
+        return this.taskList.isEmpty();
+    }
+
+    /**
+     * Adds a task to the {@code TaskList}.
+     *
+     * @param task The {@code Task} to be added to the list.
+     */
+    public void addTask(Task task) {
+        this.taskList.add(task);
+    }
+
+    /**
+     * Returns the number of tasks in the {@code TaskList}.
+     *
+     * @return The size of the task list.
+     */
+    public int getSize() {
+        return this.taskList.size();
+    }
+
+    /**
+     * Removes a task from the {@code TaskList} by its index.
+     *
+     * @param idx The index of the task to be removed (1-based index).
+     */
+    public void removeTask(int idx) {
+        Task task = this.taskList.remove(idx - 1);
+        System.out.printf("\tNoted. I've removed this task:\n\t  %s\n", task);
+    }
```
Good choice of namings. I like how all of it is simple direct yet appropriate and in accordance to the method implementation.

ahmadsyuaib reviewed (commented):
LGTM. Just some nits that you may want to consider editing. Otherwise, well done! Keep it up!

weiliann committed:
Add assertions to confirm assumptions

Assertions are not used at key points of the code execution.

The code assumes the invariance without checking and could cause a bug.

Adding assertions reduce chances of null pointer exception.

weiliann committed:
Refactor Parser class and introduce CommandFactory for better
separation of concerns

The Parser class handles both parsing user input and
instantiating command objects. The parse method is large and
handles multiple responsibilities, making the code difficult to
maintain and extend.

Combining parsing logic and command instantiation violates the
single responsibility principle. This setup makes it harder to
add new commands, as changes need to be made within the Parser
class itself, increasing the risk of bugs and reducing code
clarity.

Refactor the Parser class by splitting the parse method into
smaller, more focused methods. Introduce a new CommandFactory
class to handle the creation of command instances based on parsed
input.

This approach separates concerns by ensuring the Parser class
is responsible only for parsing raw input, while the
CommandFactory is responsible for deciding which command class
to instantiate. This separation allows for easier integration of
new commands by registering them in CommandFactory, and
handling their arguments in the Parser class.

The refactoring improves code maintainability and scalability by
simplifying the process of adding new commands and reducing the
likelihood of introducing errors when modifying parsing or
command instantiation logic.

weiliann committed:
Merge pull request #1 from weiliann/branch-A-Assertions

Add assertions to confirm assumptions

weiliann committed:
Merge assertions into code quality branch

weiliann committed:
Merge pull request #2 from weiliann/branch-A-CodeQuality

Refactor Parser class and introduce CommandFactory for better

weiliann committed:
Introduce HelpCommand to list usage for respective commands

Currently, users do not have an easy way to see the usage information
for all available commands or get specific help for a command. This
limits their ability to understand and use the application effectively.

Add a new HelpCommand feature that allows users to list usage
information for all commands or for a specific command if an argument
is provided. This provides users with a straightforward way to learn
how to use the various commands.

The HelpCommand was implemented to check for an optional argument and,
if present, display help for that specific command. If no argument is
given, it lists all available commands with their usage information.
This approach was chosen to provide flexibility while maintaining a
simple and clear user interface.

Reflection is used to dynamically retrieve the help information for
each command, making it easier to maintain and extend the help system
in the future without hard-coding each command's usage.

weiliann committed:
Merge branch 'add-help-command'

weiliann committed:
fix(storage): ensure file is created if directory does not exist

Fixed an issue in the storage class where the file was not being
saved because the required directory did not exist. The code now
checks for the existence of the directory and creates it if
necessary before saving the file.

weiliann committed:
Improve GUI design for enhanced user experience

weiliann committed:
Merge branch 'better-gui'

weiliann committed:
Add markdown for AI usage

weiliann committed:
Add user guide for joe bot
